### PR TITLE
Redis

### PR DESCRIPTION
**Pull request by p-alik**
PR aims to fix the issue [Implement priorities for redis queues](https://github.com/gearman/gearmand/issues/61).
To put multiple items (workload and priority) into redis storage [redis set command](https://redis.io/commands/set) used in original implementation is replaced by [hmset](https://redis.io/commands/hmset).

### CODE REVIEW

**Comment by SpamapS** [comment]
recheck

**Comment by SpamapS** [comment]
I have one concern. If a user were to shutdown a redis backed gearmand, upgrade to this version, and start up, would the \0 terminated null keys already in the database be handled properly? Or are the \0 terminations not actually happening in the database?

**Comment by p-alik** [comment]
gearmand <= v1.1.15 queues into redis with keys terminated by \0.
After update/reboot of gearmand `_hiredis_replay` should restore the queue without bothering about \0. `_hiredis_done` wouldn't be able to delete restored job after it's done by some worker. So the job remains in Redis.

**Comment by SpamapS** [comment]
Hm, so could we add a bit that migrates the old keys on startup? Should be relatively quick, and could be disabled for those who want a faster startup if they aren't updating.

**Comment by p-alik** [comment]
I'll see how to do that if you sure it is a good idea.

**Comment by SpamapS** [comment]
If it turns out to be really hard we can poll the mailing list. I'm just concerned that we'd break somebody's system with an update.

**Comment by p-alik** [comment]
I'll try to do that.

**Comment by jamielennox** [comment]
recheck

**Comment by p-alik** [comment]
please check

**Comment by SpamapS** [comment]
Hey @p-alik .. We're transitioning BonnyCI to an integration from webhooks.. so it may be disconnected at the moment. I'll hook it back up before merging. Thanks for working on the backward compat mode, I'll review soon.

**Comment by SpamapS** [comment]
recheck

File: libgearman-server/plugins/queue/redis/queue.cc [hunk 1]
```diff
@@ -124,16 +124,39 @@ bool gearmand::plugins::queue::Hiredis::hmset(vchar_t key, const void *data, siz
...
   if (reply == nullptr)
     return false;
 
-  // 2 x (key + value)
-  assert(reply->elements == 4);
+  //FIXME remove workaround
+  if(reply->type == REDIS_REPLY_ERROR) {
+    // workaround to ensure gearmand upgrade.
+    // gearmand <=1.1.15 stores data in string, not in hash.
+    gearmand_log_warning(GEARMAN_DEFAULT_LOG_PARAM, "redis replies for HGETALL: %s", reply->str);
```
Comment (SpamapS) [inline libgearman-server/plugins/queue/redis/queue.cc]: Is this really a warning? The only time we fetch from the DB is on startup. And this is entirely expected behavior the first time one starts an upgraded gearmand. Seems more like INFO to me.

**Pull request merged**